### PR TITLE
Improve errors when user attempts to save unwritable data to VCF

### PR DESCRIPTION
**Pull request by henrydavidge**
## What changes are proposed in this pull request?
Previously, if a user tried to a write a DataFrame that was missing required fields like `contigName`, they could get confusing errors that they were trying to write an empty DF. Since our default stringency is "silent" (perhaps a bad decision!), we'd drop VCs if they failed conversion. Now, before writing we check the schema to make sure that required fields are present. In addition, I cleaned up the behavior for "strict" stringency so that conversion errors cause the write to fail.

## How is this patch tested?
- [x] Unit tests
- [ ] Integration tests
- [ ] Manual tests

(Details)

### CODE REVIEW

File: core/src/main/scala/io/projectglow/vcf/VCFFileFormat.scala [hunk 1]
```diff
@@ -273,6 +273,23 @@ object VCFFileFormat {
       false
     }
   }
+
+  def requireWritableAsVCF(schema: StructType): Unit = {
+    val baseRequiredFields =
+      Seq(VariantSchemas.contigNameField, VariantSchemas.startField, VariantSchemas.refAlleleField)
```
Comment (karenfeng) [inline core/src/main/scala/io/projectglow/vcf/VCFFileFormat.scala:279]: I believe `endField` is also required as well. To check this more thoroughly, we could have tests that remove each of the columns in a DataFrame and attempt conversion to a `VariantContext`. If conversion fails, we should verify that the check for writable-ness also fails, and vice versa.

File: core/src/main/scala/io/projectglow/vcf/VCFFileFormat.scala [hunk 2]
```diff
@@ -273,6 +273,23 @@ object VCFFileFormat {
...
+    val requiredFields = if (schema.exists(_.name == VariantSchemas.genotypesFieldName)) {
+      baseRequiredFields :+ VariantSchemas.alternateAllelesField
```
Comment (karenfeng) [inline core/src/main/scala/io/projectglow/vcf/VCFFileFormat.scala:281]: Is this strictly true? It looks like we can successfully make `Genotype`s as long as the appropriate allele exists in the `VariantContext` (i.e. only called reference allele).

File: core/src/test/scala/io/projectglow/vcf/VCFFileWriterSuite.scala [hunk 3]
```diff
@@ -400,6 +400,31 @@ abstract class VCFFileWriterSuite(val sourceName: String)
     assert(sampleIds.length == 1)
     assert(sampleIds.head == Seq("sample_1", "sample_2", "sample_3"))
   }
+
+  test("validate schema before write") {
```
Comment (karenfeng) [inline core/src/test/scala/io/projectglow/vcf/VCFFileWriterSuite.scala:404]: This would benefit from more thorough testing - at minimum, should iterate through each of the expected required fields. Ideally, we should have a sanity check to make sure that our expected required fields match the actual required fields.

File: core/src/main/scala/io/projectglow/vcf/InternalRowToVariantContextConverter.scala [hunk 4]
```diff
@@ -166,7 +166,9 @@ class InternalRowToVariantContextConverter(
       try {
         Option(builder.make())
       } catch {
-        case NonFatal(ex) =>
+        case NonFatal(_) if stringency == ValidationStringency.SILENT =>
+          None
+        case NonFatal(ex) if stringency == ValidationStringency.LENIENT =>
```
Comment (karenfeng) [inline core/src/main/scala/io/projectglow/vcf/InternalRowToVariantContextConverter.scala:171]: I believe we already do this behavior with `provideWarning`.

File: core/src/main/scala/io/projectglow/vcf/VCFFileFormat.scala (hunk 2 above)
Comment (henrydavidge) [reply core/src/main/scala/io/projectglow/vcf/VCFFileFormat.scala:286]: As discussed today, I think it's better to disallow this. It's possible to form a variant context, but unlikely to result in the behavior users expect. They can always add a dummy `alternateAlleles` if it's really what they want.

File: core/src/test/scala/io/projectglow/vcf/VCFFileWriterSuite.scala (hunk 3 above)
Comment (henrydavidge) [reply core/src/test/scala/io/projectglow/vcf/VCFFileWriterSuite.scala:417]: Done
